Design doc copy to s3

moulimukherjee opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->
Rendered version: https://github.com/moulimukherjee/materialize/blob/design-doc-copy-to-s3/doc/developer/design/20231117_copy_to_s3.md

### Motivation
We want to support a one time export of a snapshot to s3 https://github.com/MaterializeInc/materialize/issues/7256

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+- If the S3 path given already has some data in it, we should fail the operation.
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows and
```
Yes, Snowflake does, which I thought was nice https://docs.snowflake.com/en/sql-reference/sql/copy-into-location#copy-options-copyoptions. I will add the link to the design doc as well.

They show the number of rows and the amount of data. If `DETAILED_OUTPUT` flag is set, then they also display the list of files and their details.

moulimukherjee replied:
Oh and postgres also returns the number of rows https://www.postgresql.org/docs/current/sql-copy.html

benesch replied:
> Oh and postgres also returns the number of rows https://www.postgresql.org/docs/current/sql-copy.html
> 
> > `COPY count`
> > The count is the number of rows copied.

Yep—but I'm worried that we don't have further flexibility, and the _only_ thing we can return from a `COPY` operation is the number of rows copied.

moulimukherjee replied:
I overlooked that Snowflake query is not exactly postgres sql. It makes sense for us to only return number of rows. Will update.

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
```
Yes, I agree. I kept the Storage one as a rejected alternative.

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
````diff
@@ -0,0 +1,129 @@
+
+```sql
+COPY <mv_name> -- mv or table name
+TO s3://prefix/path/
````
Yes, I had not updated since the discussion. Will change this 👍

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+### SQL
+In materialize we already have some support for `COPY`
+commands. We should expand upon them, keeping them as close to
+the [postgres syntax](https://www.postgresql.org/docs/current/sql-copy.html).
```
👍

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+into S3 is a simple and intuitive way to enable this for our customers.
+
+## Goals
+- Be able to run a sql command to write a snapshot of a materialized view or a table (persist backed items) to S3.
+- Initially the scope will be limited to only parquet as an output format.
+- Users should be able to get a response back indicating some information about the data written to S3, for e.g., number of rows, bytes transferred etc.
```
Ack. I will rephrase.

moulimukherjee replied:
Note: Since I will be out most of December, I am changing "end of Q4" to "early Q1 next year".

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+- Users should be able to get a response back indicating some information about the data written to S3, for e.g., number of rows, bytes transferred etc.
+
+## Non-goals
+- Support arbitrary select queries like `COPY (SELECT ...) TO S3` in the initial version.
```
👍

benesch commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.Everything which runs, like sources or sinks, do so continuously once scheduled.
+
+### Output file format
+We should start with a single file format and add more if there's more user interest. Currently,
+there has been request for parquet, so that will be the initial supported format.
```
If we go with the compute implementation, I'm not sure that it will be more work to support arbitrary queries! In fact, it might actually be harder to _disallow_ arbitrary queries.

moulimukherjee replied:
I had tried some changes locally to see how things will fit. It seemed like it would be fairly easy to only support tables/mvs? Basically we'll have to do something similar to https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/optimize/subscribe.rs#L164-L194 (and not do the other match arm which is for a query).

moulimukherjee replied:
I updated to start with csv format first cc @hlburak

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,129 @@
+USING AWS CONNECTION aws_conn -- previously created aws connection
+WITH (
+  FORMAT = 'parquet', -- only parquet support for the time being
+  MAX FILE SIZE = '2g' -- config to limit file sizes in s3
```
Yes, this is on my list to look into.

moulimukherjee replied:
(Will update in the doc)

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+version we should stick to the default values as per the
+[Postgres COPY command](https://www.postgresql.org/docs/current/sql-copy.html).
+
+### Bounded Memory
+I had wrongly assumed that with reading directly from persist we could use backpressure similar to what we do in upsert. As @guswynn correctly pointed out, that wouldn't work due to the following:
+* For upsert, we read configured amount of data from persist, but we eventually consolidate in rocksdb (or in memory if disk is not enabled) which maintains the upsert state.
+* In compute, we may be able to fetch data in chunks from persist, but it will still need to be
+consolidated for the entire shard and this will have to happen in memory.
+
+One possibility could be that we actually use disk (with rocksdb) in Compute to keep state similar
+to upsert and consolidate the data there before writing the snapshot to S3. This seems complicated
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
+
+Since backpressure is not feasible right off the bat when reading from materialized view as I had
+assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
+doing the [multi part uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
+to S3.
+* A single worker starts a multi part upload with a `CreateMultipartUpload` API call to start
+writing data to a S3 file in the given path.
+* The worker should keep a counter, say `bytes_to_write` to keep track of the aggregate bytes it will write.
+* The worker should also keep some state to track the number of parts in the upload
+and the corresponding etags returned from AWS for the uploaded parts.
+* The worker receives data to be written to S3, adds it to `bytes_to_write` to see if it will exceed
+MAX FILE SIZE.
+* If less than MAX FILE SIZE, the worker assigns an incremental part number starting from 1,
+makes a `UploadPart` call and keeps the returned `ETag` along with the part number in memory.
+* If more than MAX FILE SIZE, the worker makes a `CompleteMultipartUpload` call to finish
+the upload of a single file and provides the part number and etags in the request.
+* Then the worker creates a new multi part request to write to a new S3, resets the
+`bytes_to_write`, clears the part number with etag mapping in memory and uploads the part for
+the new multi part call.
```
@guswynn Does this make sense?

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+s3 prefixes so that we can write the data.
+  - `s3:ListBucket`: We need this permission to make sure that we are writing to an empty path.
+  - `s3:PutObject`: This is required to actually upload files to S3.
+- If they try to run a `COPY` to s3 without all the permission, then the operation should fail
```
```suggestion
- If they try to run a `COPY` to S3 without all the permissions, then the operation should fail
```

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+
+#### S3 Path
+For the s3 path, users should be able to provide an expression
+like `'s3://prefix/path/' || mz_now()` to help generate the path. This would
```
How is the path determined when multiple files are written (e.g. if the max size is set to a value that is less than the size of the table)?

Is this path a prefix and we have some known format for the per-file suffix?

moulimukherjee replied:
Ah, this s3 path should be an empty directory i.e. just the path prefix. I added a section for s3 file names below, will link here and further clarify.

danhhz commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
```
Compute tries pretty hard to evenly balance work between workers, so if this ends up being run in compute, I'm not sure this is the place to cut scope for a v1 :). (I also don't think it's terribly hard to make it multi worker? Each worker could be doing the logic you mention below but pointed at some `<prefix>/<worker_idx>/part-<idx>.csv`)

moulimukherjee replied:
Instead of the path, maybe we can add the worker_id to the filename itself, something like `<prefix>/part-<worker_idx>-<idx>.csv`. I am not sure how would this affect ordering though (there's an open question around supporting ORDER BY).

The reason being, eventually we want the data exported to S3 be used in a data warehouse like Snowflake most likely via external tables. In that scenario, all the data files should be present at the given prefix. We could potentially make it part of the path like a partition `<prefix>/worker_id=<worker_idx>/part-<idx>.csv`, and import `<prefix>` as the table, but this adds the worker ID partition column unnecessarily. There might be other ways to read sub folders, but they won't work right out of the box.

moulimukherjee replied:
While copying to S3, if the given path is not empty, the command is going to fail (similar to Snowflake). Also, we are going to support scalar expressions in s3 path. I would imagine for most of the use cases the path is going to be something like `'s3://some/path' + mz_now_or_some_date_time_suffix`. So, I would imagine there's very less chance of conflict happening but it's definitely not impossible. There's a window between the COPY command and when the first file is written. Hmm...maybe we should write an in progress placeholder?

But I do like the idea of writing a manifest file (we used to use versions of this widely at Stripe for various consistency checks). Snowflake does not, but Redshift does have a concept of manifest file to load data from S3 https://docs.aws.amazon.com/redshift/latest/dg/loading-data-files-using-manifest.html.

I would prefer if the folder structure like below, i.e. not nested under batches so that the data is exactly under the user specified path (and they can use it to load it into data warehouses predictably).
```
<prefix>
│── aaaa.csv
│── bbbb.csv
│── cccc.csv
└── progress_[0]-[<as_of>+1].json
```
I will have to check though if Snowflake can successfully ignore separate manifest files.

moulimukherjee replied:
Snowflake does support `PATTERN = 'regex_pattern'` in their `CREATE EXTERNAL TABLE` which can be used to only include files of a certain pattern.

benesch replied:
I can see pros and cons to each. No strong feelings myself. Is this the only open question? If so we could potentially mark it down as an open question that we'll resolve later during the implementation.

moulimukherjee replied:
@benesch Yes, this is the only one. Sounds good, adding it to open questions.

moulimukherjee replied:
Added to open questions

hlburak commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+### More observability
+Postgres reports the progress of an ongoing `COPY` to a `pg_stat_progress_copy`. We can have a
+similar table to show progress which the user can query to see how far along they are.
+
```
Worth also calling out:

* *Partitions:* Since the target use case here is unloading data to an analytical data warehouse downstream, users will probably ask about partitions sooner or later for query performance.

* *Scheduling:* Most users will have an orchestrator or manage custom scripts for executing jobs like these, but certainly some will want to set a copy out schedule in Materialize. We should support this.

moulimukherjee replied:
TBH I am unsure about Scheduling and if it's something we should support in Materialize, but I am adding `Partitions` as potential future work item.

hlburak commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+Initially supporting `MAX FILE SIZE` as integer for the bytes will be pretty simple. Later we
+can introduce a `Memory` type which can take either an integer or a parseable string like `'2 GB'`,
+similar to the implementation for `Duration`.
+
```
Are there performance benefits to selecting a certain default max file size? Snowflake has some [recommended file sizes](https://docs.snowflake.com/en/user-guide/tables-external-intro) for reading data via external tables which we can use to choose a default.

moulimukherjee replied:
Not much while actually writing the data. The performance benefit is on the querying data warehouse side. Usually it's faster to query large files instead of smaller files.

moulimukherjee replied:
Snowflake recommends 16-256MB for all other data types including csv for performance, so 256MB can be our max file size.

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+Everything which runs, like sources or sinks, do so continuously once scheduled till they are
+dropped.
+
+### Output file format
```
Maybe for pretty simple datasets it should be fine. In general I don't think CSV can be lossless or at least it will need a lot of tweaks. It will also depend entirely upon what the options are when exporting and while importing. For example if somebody wants NULLs as empty string, it's changing the data type. Do we map strings to varchar or text?

Also, I need to check what happens with nested objects. I am assuming it will get serialized to some string, not sure if it will have enough information for it to be deserialized correctly.

moulimukherjee replied:
Btw, I just tried reimporting an exported nested type in postgres without overriding any option, it failed with invalid syntax because of nested double quotes. Will try later if it's possible to make it work with tweaks.

moulimukherjee replied:
Updated that it can be lossless if we escape correctly

petrosagg commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
```
> That does mean we should be careful with buffers and such in the implementation of COPY TO S3—we don't want the implementation to introduce any new uses of memory.

Cool, that leaves arrangements as the only option because we still need to do consolidation and arrangements is the only way to not introduce new uses of memory.

benesch replied:
Oh, sorry, I think it's fine to stick a `.consolidate()` operator where you need it, while `persist_source` doesn't do consolidation on read. I just want to make sure that once persist supports CoR, it's ~no work to make `COPY persist_backed_relation TO S3` use bounded memory.

moulimukherjee replied:
Updated section to mention that we should be careful with memory, and we are able to leverage persist consolidate on read when available.

petrosagg commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,279 @@
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
+doing the [multi part uploads](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)
+to S3.
+* A single worker starts a multi part upload with a `CreateMultipartUpload` API call to start
```
I'm assuming there is a hidden parameter here which defines the maximum amount of data that we're willing to store in memory before creating a part which might be less than the max file size, right? AWS suggests that this should be around the 100MB mark so maybe we can spell it out in the design doc.

moulimukherjee replied:
Makes sense, will add.

moulimukherjee replied:
Added.

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
````diff
@@ -0,0 +1,129 @@
+```sql
+COPY <mv_name> -- mv or table name
+TO s3://prefix/path/
+USING AWS CONNECTION aws_conn -- previously created aws connection
````
Yup, that's on my list to get feedback from SQL council. Will ping in the channel after addressing the comments.

benesch replied:
> The aws connection seemed specific to using s3 as a target whereas the `WITH` block is more about the data (eg format) itself? I would defer to you here though.

Yeah, I'd just stuff it into the `WITH` block. The `USING SCHEMA REGISTRY` clause of `CREATE SOURCE` is something I regret, for example, and I wish we'd instead just made `SCHEMA REGISTRY` a parameter of the format options (`FORMAT AVRO (SCHEMA REGISTRY = ...)`).

> I think that might be worth doing now that we're supporting arbitrary queries. We want something that feels natural whether the user is copying over an existing object or using a query.

Between PostgreSQL's `COPY TO`, Snowflake's `COPY INTO`, and Redshift's `UNLOAD`, the syntax for specifying the copy source as either a query or a table seems well established. Unless there's something I'm missing, I don't think there's anything to workshop on that specific point.

moulimukherjee replied:
> The USING SCHEMA REGISTRY clause of CREATE SOURCE is something I regret...

Ah makes sense. I was sort of using it as inspiration.

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
````diff
@@ -0,0 +1,303 @@
+```
+
+Note: We'll not support `ORDER BY` clause in the select query because it will be
+very expensive. As a workeraound when users are using the data as an external
````
```suggestion
very expensive. As a workaround when users are using the data as an external
```

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,303 @@
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+We should use the Compute clsuter. Compute already has a concept of
```
```suggestion
We should use the Compute cluster. Compute already has a concept of
```

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,303 @@
+
+Also, since backpressure is not feasible right off the bat when reading from materialized view as
+I had assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
```
does this mean the initial scope of this project will include support for arbitrary SELECT queries (besides those using `ORDER BY`)? A few statements earlier in this design doc indicate that SELECT will be a follow-on feature, with just MV/table support to start.

moulimukherjee replied:
Yes, we would support SELECT queries. Corrected the tense at a couple of remaining places.

rjobanp commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,303 @@
+similar to a [Postgres `COPY` output](https://www.postgresql.org/docs/current/sql-copy.html) of
+`COPY <row_count>`.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
```
If a user incorrectly handles or loses the result of the `COPY` command, there is a chance they won't know if the path they have in S3 is incomplete or not.

Would it make sense to introduce some sort of sentinel file to indicate a complete copy has occurred? Either something we write at the beginning (e.g. an empty blob named `INCOMPLETE`) that is deleted at the end of the copy, or a `COMPLETE` blob written at the end to indicate full completion? The latter would be easier since it wouldn't require the DeleteObject permission.

This wouldn't be that critical, but might help simplify confusion when errors do occur.

moulimukherjee replied:
Yeah, there's an open question around having a manifest file at the end https://github.com/MaterializeInc/materialize/pull/23556/files#diff-e90b6157a78c239d2480f26f26cd737f66ea6813bd1fc149294b23b9f76d0e68R297 which has the same purpose. IMO it's better to write a file at the end instead of removing, exactly as you mentioned it would not require DeleteObject.

rjobanp reviewed (commented):
nice work!!

benesch commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,304 @@
+#### Testing after code is merged
+- Switch on the feature flag for the staging environment and do an end to end copy to s3
+flow in staging.
+- Also, try out a scenario where we cancel the copy midway
```
We should also add a benchmark (not sure if a feature benchmark or a scalability benchmark here is more appropriate) that copies out a material amount of data and measures how long it takes.

benesch reviewed (approved):
LGTM! I think you'll want @petrosagg to weigh in as well before merging.

antiguru commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,304 @@
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+We should use the Compute cluster. Compute already has a concept of
```
Please keep in mind that there may be multiple replicas running the same sink, so it'll need to coordinate accordingly. Also, compute doesn't currently assume that replicas run to completion, so the sink must tolerate being stopped and recreated at any time.

An alternative we should consider is that we create an ephemeral materialized view based on the select expression, und do the actual uploading in storage.

benesch replied:
> Please keep in mind that there may be multiple replicas running the same sink, so it'll need to coordinate accordingly. Also, compute doesn't currently assume that replicas run to completion, so the sink must tolerate being stopped and recreated at any time.

I think it should be fine as long as the computation is perfectly deterministic, right? Then no coordination is required. 

> An alternative we should consider is that we create an ephemeral materialized view based on the select expression, und do the actual uploading in storage.

Are there benefits to this approach? Storage will one day soon support replication.

antiguru replied:
> I think it should be fine as long as the computation is perfectly deterministic, right? Then no coordination is required.

Not if the worker ID is part of the file names. In this case, the output from different replicas is equivalent, but the sharding is different.

> Are there benefits to this approach? Storage will one day soon support replication.

None other than different splitting of concerns. I have no strong feeling here!

benesch replied:
> Not if the worker ID is part of the file names. In this case, the output from different replicas is equivalent, but the sharding is different.

Oh, that's a really good point. Hrm. We'll need to sort that out.

moulimukherjee replied:
It's not going to be as efficient, but we could start with only a single worker writing to s3.

Also, when a compute sink is killed, there could be multiple multi part uploads in progress. Initially I was thinking these could be kept worker local state, but that wouldn't work if the replicas are themselves killed. What would be a good way to store the progress? Would it need to be similar to a progress collection in storage?

antiguru replied:
There's a certain risk involved with moving everything to a single worker, which is that it might stall out the rest of the system. We should probably discuss alternatives that either don't have this property or avoid it.

antiguru replied:
@benesch and I briefly talked about this, and there seems to be a way out of this:
* Pick a number of parts, and tell the sink how many parts it should have.
* The number could have a default, say 32, or whatever the largest replica is at creation time, or it could be specified by the user.
* Assign records based on a hash function to parts, exchange the data based on the part to workers. We can workshop how this works in detail.
* Upload the individual parts, name them by their part ID.

To produce deterministic outputs, consider:
* adding a read hold for the timestamp to be copied, which corresponds to the as-of of the compute dataflow.
* Then, restarts would use the same timestamp and since it's deterministic, upload the same data again.
* There's probably some work required to teach the adapter about copy commands, and they need to be in the catalog so they can survive envd restarts.

moulimukherjee replied:
Thanks, I will add this to the doc.

teskje replied:
> The number could have a default, say 32, or whatever the largest replica is at creation time, or it could be specified by the user.

I'd say number of workers in the largest replica at creation time is a good heuristic. I bet the number of cases where someone starts a `COPY` command and then proceeds to change the replica configuration while that is running will be exceedingly rare, so we don't have to optimize for this case.

> There's probably some work required to teach the adapter about copy commands, and they need to be in the catalog so they can survive envd restarts.

Why do we need them to survive envd restarts? I don't think it's clearly specified in this design but AIUI the idea it that a `COPY TO S3` command always only copies a point-in-time snapshot of whatever it reads from. If this is true then it behaves like a `SELECT` in that we can forget about it on client disconnects or envd restarts.

If my understanding is not correct and we do want to support continuous copying of collections to S3 somehow, I think we'd run into problems with determinism. Keep in mind that between envd restarts we could have deployed a new release, which might have changed the optimizer or rendering so that we now produce different outputs than before.

moulimukherjee replied:
> Arguably, one shot operations should be a CREATE operation that creates a named catalog object...

I will add this in future work.

teskje commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,304 @@
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
```
A thing that's missing in this discussion is that we will need to invent some new way to communicate possible sink errors, like the ones described in "User Experience" above. So far all existing sinks communicate errors by writing them into the sinked collection, but that won't be possible in this case if we want to show errors back to the user via pgwire.

moulimukherjee replied:
OOC what happens in case of SELECTs or SUBSCRIBEs?

teskje replied:
Both communicate results back to envd using `ComputeResponse`s (either `PeekResponse` or `SubscribeResponse`). And both response types have an `Error` variant that gets sent when an error is encountered along the way. Then adapter has some dedicated code to handle these error responses by converting them into pgwire errors (and, in the case of `SUBSCRIBE`s, cancelling them).

benesch replied:
> invent some new way to communicate possible sink errors, like the ones described in "User Experience" above

Any suggestions on what that protocol would look like, @teskje?

moulimukherjee replied:
Ah so similar to PeekResponse at https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/protocol/response.rs#L210-L217, we can have something like 

```
pub enum CopyResponse {
    /// Returned number of rows copies.
    Rows(u64),
    /// Error of an unsuccessful copy.
    Error(String),
    /// The copy was canceled.
    Canceled,
}
```

teskje replied:
I haven't thought about this in depth, but perhaps we should introduce a new response type that resembles `PeekResponse` and has an `Ok` and an `Error` variant, to communicate the result of the copy operation. Or, if we wanted to return intermediary status updates, something like a `SubscribeResponse` would work, that continually produces updates about, e.g. the number of bytes uploaded.

moulimukherjee replied:
Added `CopyResponse` section https://github.com/moulimukherjee/materialize/blob/design-doc-copy-to-s3/doc/developer/design/20231117_copy_to_s3.md#copyresponse

teskje commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,304 @@
+
+One possibility could be that we actually use disk (with rocksdb) in Compute to keep state similar
+to upsert and consolidate the data there before writing the snapshot to S3. This seems complicated
+and I don't think it will be worth the effort. Also, given that we might
```
```suggestion
and I don't think it will be worth the effort. Also, given that we will
```

I think we're pretty sure about that feature at this point :)

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
```diff
@@ -0,0 +1,304 @@
+["Compute Sinks"](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/compute-types/src/sinks.rs#L100-L103),
+which include Subscribe and Persist (for materialized views). We can create another
+Compute S3 sink, which will actually read from persist and write to S3. Keeping this in a
+compute cluster also means,
```
Ah, my mistake. The scope was initially limited to only support materialized views and later SELECT queries were added. I will update.

moulimukherjee commented on doc/developer/design/20231117_copy_to_s3.md:
````diff
@@ -0,0 +1,304 @@
+  FORMAT = 'csv', -- file format name
+  MAX FILE SIZE = 16777216 -- max file size in bytes, to limit size of files in s3
+)
+```
````
Since this is meant to be a one shot dumping of data to S3 via the copy command, it should not.

teskje commented on doc/developer/design/20231117_copy_to_s3.md:
````diff
@@ -0,0 +1,305 @@
+The SQL would look like the following.
+
+```sql
+COPY <mv_name or select_query> -- mv or table name or a valid SELECT query
````
Is there a reason why this is limited to MVs or tables? Why can't we copy the contents of, e.g., a view or a source?

moulimukherjee replied:
Ah yes. Technically it's any relation + SELECT queries. Updating.

moulimukherjee commented:
cc @petrosagg @MaterializeInc/compute @antiguru Have updated the design doc and added a section for "Writing to S3 deterministically" and addressed the other comments. Lmk if anything looks incorrect.

petrosagg reviewed (approved):
This design feels like it's in a good enough place for beginning the implementation :tada:. If things come up that need to change the design for we can do it with new PR. Thanks for addressing all these comments!
